samples/csharp_dotnetcore/12.customQABot/Bots/CustomQABot.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Builder.AI.QnA;
+using Microsoft.Bot.Builder.AI.QnA.Models;
+using Microsoft.Bot.Schema;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.BotBuilderSamples.Bots
+{
+    public class CustomQABot : ActivityHandler
+    {
+        private readonly ILogger<CustomQABot> _logger;
+        private readonly IHttpClientFactory _httpClientFactory;
+        private readonly string _endpointKey;
+        private readonly string _hostname;
+        private readonly string _knowledgeBaseId;
+        private readonly string _defaultWelcome = "Hello and Welcome";
+        private readonly bool _enablePreciseAnswer;
+        private readonly bool _displayPreciseAnswerOnly;
+
+        public CustomQABot(IHttpClientFactory httpClientFactory, IConfiguration configuration, ILogger<CustomQABot> logger)
+        {
+            _logger = logger;
+            _httpClientFactory = httpClientFactory;
+            const string ErrorMessage = "Configuration parameter is missing or empty.";
+
+            _hostname = configuration["LanguageEndpointHostName"];
+            if (string.IsNullOrEmpty(_hostname))
+            {
+                throw new ArgumentException(ErrorMessage, "LanguageEndpointHostName");
+            }
+
+            _endpointKey = configuration["LanguageEndpointKey"];
+            if (string.IsNullOrEmpty(_endpointKey))
+            {
+                throw new ArgumentException(ErrorMessage, "LanguageEndpointKey");
+            }
+
+            _knowledgeBaseId = configuration["ProjectName"];
+            if (string.IsNullOrEmpty(_knowledgeBaseId))
+            {
+                throw new ArgumentException(ErrorMessage, "ProjectName");
+            }
```

Comment:
How about this? The message reads better, I am also dropping the argument name because "LanguageEndpointKey" and the others are not arguments, they are values in the configuration parameter that got received. 
```suggestion
            const string missingConfigError = "{0} is missing or empty in configuration.";
            _hostname = configuration["LanguageEndpointHostName"];
            if (string.IsNullOrEmpty(_hostname))
            {
                throw new ArgumentException(string.Format(missingConfigError, "LanguageEndpointHostName"));
            }

            _endpointKey = configuration["LanguageEndpointKey"];
            if (string.IsNullOrEmpty(_endpointKey))
            {
                throw new ArgumentException(string.Format(missingConfigError, "LanguageEndpointKey"));
            }

            _knowledgeBaseId = configuration["ProjectName"];
            if (string.IsNullOrEmpty(_knowledgeBaseId))
            {
                throw new ArgumentException(string.Format(missingConfigError, "ProjectName"));
            }
```